domain/upgrade/state/state.go

```diff
@@ -172,198 +181,163 @@ func (st *State) StartUpgrade(ctx context.Context, upgradeUUID domainupgrade.UUI
-	lookForDoneNodesStmt, err := st.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
+WHERE  upgrade_info_uuid = $ControllerNodeInfo.upgrade_info_uuid
+AND    controller_node_id = $ControllerNodeInfo.controller_node_id;
+`, controllerNodeInfo)
 	if err != nil {
-		return errors.Annotatef(err, "preparing select done query")
+		return errors.Annotate(err, "preparing select done query")
 	}
 
-	setNodeToDoneQuery := `
+	setNodeToDoneStmt, err := st.Prepare(`
 UPDATE  upgrade_info_controller_node
 SET     node_upgrade_completed_at = DATETIME("now")
-WHERE   upgrade_info_uuid = $M.info_uuid
-AND     controller_node_id = $M.controller_id
+WHERE   upgrade_info_uuid = $ControllerNodeInfo.upgrade_info_uuid
+AND     controller_node_id = $ControllerNodeInfo.controller_node_id
 AND     node_upgrade_completed_at IS NULL;
-`
-	setNodeToDoneStmt, err := st.Prepare(setNodeToDoneQuery, sqlair.M{})
+`, controllerNodeInfo)
 	if err != nil {
 		return errors.Annotatef(err, "preparing update node query")
 	}
 
-	completeUpgradeQuery := `
+	m := sqlair.M{
+		"from_state": upgrade.DBCompleted,
+		"to_state":   upgrade.StepsCompleted,
+	}
+	completeUpgradeStmt, err := st.Prepare(`
 UPDATE upgrade_info
 SET    state_type_id = $M.to_state
-WHERE  uuid = $M.info_uuid AND state_type_id = $M.from_state
+WHERE  uuid = $Info.uuid AND state_type_id = $M.from_state
 AND (
     SELECT COUNT(*)
 	FROM   upgrade_info_controller_node
-    WHERE  upgrade_info_uuid = $M.info_uuid
+    WHERE  upgrade_info_uuid = $Info.uuid
     AND    node_upgrade_completed_at IS NOT NULL
 ) = (
     SELECT COUNT(*) 
 	FROM   upgrade_info_controller_node
-    WHERE  upgrade_info_uuid = $M.info_uuid
+    WHERE  upgrade_info_uuid = $Info.uuid
 );
-`
-	completeUpgradeStmt, err := st.Prepare(completeUpgradeQuery, sqlair.M{})
+`, info, m)
 	if err != nil {
 		return errors.Annotatef(err, "preparing complete upgrade query")
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		var node infoControllerNode
-		err := tx.Query(ctx, lookForDoneNodesStmt, sqlair.M{
-			"info_uuid":     upgradeUUID,
-			"controller_id": controllerID,
-		}).Get(&node)
+		var node ControllerNodeInfo
+		err := tx.Query(ctx, lookForDoneNodesStmt, controllerNodeInfo).Get(&node)
 		if err != nil {
 			if errors.Is(err, sqlair.ErrNoRows) {
 				return errors.Errorf("controller node %q not ready", controllerID)
 			}
 			return errors.Trace(err)
 		}
 
-		err = tx.Query(ctx, setNodeToDoneStmt, sqlair.M{
-			"info_uuid":     upgradeUUID,
-			"controller_id": controllerID,
-		}).Run()
+		err = tx.Query(ctx, setNodeToDoneStmt, controllerNodeInfo).Run()
 		if err != nil {
 			return errors.Trace(err)
 		}
 
 		var outcome sqlair.Outcome
-		err = tx.Query(ctx, completeUpgradeStmt, sqlair.M{
-			"info_uuid":  upgradeUUID,
-			"from_state": upgrade.DBCompleted,
-			"to_state":   upgrade.StepsCompleted,
-		}).Get(&outcome)
+		err = tx.Query(ctx, completeUpgradeStmt, info, m).Get(&outcome)
```

Comment:
Unused `outcome`

Reply:
Ah yea, looks like this should just be removed. This query sets the state to done if the number of completed controllers is equal to the number of controllers. I.e. for any particular one you don't know if its going to be the last controller node to finish so don't know how many rows to expect here.